flowz/test/artifacts/artifacts_test.py

```diff
@@ -0,0 +1,295 @@
+
+        tools.assert_false(artifact.exists())
+
+        value = yield artifact.get()
+
+        tools.assert_equal(value, 'one')
+        tools.assert_true(artifact.exists())
+        tools.assert_true(artifact.ensure())
+
+    @tt.gen_test
+    def test_keyed_artifact_transform(self):
+        def deriver(key, num_dict):
+            return num_dict[key]
+
+        def rderiver(num_dict, val):
+            for (k, v) in num_dict.iteritems():
+                if v == val:
+                    return k
+            return None
+
+        key = 1
+        artifact = KeyedArtifact(key, DerivedArtifact(deriver, key, self.NUM_DICT))
+        artifact2 = artifact.transform(rderiver, self.NUM_DICT)
+
+        key2 = yield artifact2.get()
+        tools.assert_equal(key, key2)
+
+    @tt.gen_test
+    def test_keyed_artifact_threaded_transform(self):
+        def deriver(key, num_dict):
+            return num_dict[key]
+
+        def rderiver(num_dict, val):
+            for (k, v) in num_dict.iteritems():
+                if v == val:
+                    return k
+            return None
+
+        executor = futures.ThreadPoolExecutor(1)
+
+        key = 1
+        artifact = KeyedArtifact(key, DerivedArtifact(deriver, key, self.NUM_DICT))
+        artifact2 = artifact.threaded_transform(executor, rderiver, self.NUM_DICT)
+
+        key2 = yield artifact2.get()
+        tools.assert_equal(key, key2)
+
+    @tt.gen_test
+    def test_maybe_artifact(self):
+        def deriver(key, num_dict):
+            return num_dict[key]
+
+        # prove that both artifacts and non-artifacts result in futures
+        key = 1
+        artifact = DerivedArtifact(deriver, key, self.NUM_DICT)
+        future1 = maybe_artifact(artifact)
+        tools.assert_is_instance(future1, tornado.concurrent.Future)
+
+        future2 = maybe_artifact('one')
+        tools.assert_is_instance(future2, tornado.concurrent.Future)
+
+        val1 = yield future1
+        val2 = yield future2
+        tools.assert_equal(val1, val2)
+
+        # Make sure that just having a "get" function isn't enough to be an artifact!
+        dict_ = {1: 'one'}
+        tools.assert_true(hasattr(dict_, 'get'))
+        future3 = maybe_artifact(dict_)
+        val3 = yield future3
+        tools.assert_equal(val3, dict_)
+
+    # ****** End with helper methods on AbstractArtifact ******
+
+    @tt.gen_test
+    def test_as_channel(self):
+        @gen.coroutine
+        def getter():
+            raise gen.Return('foo')
+
+        artifact = ExtantArtifact(getter)
+        channel = artifact.as_channel()
+        result = yield channel.start()
+        tools.assert_true(result)
+
+        artifact2 = yield channel.next()
+        tools.assert_equal(artifact, artifact2)
+
+        val = yield artifact2.get()
+        tools.assert_equal(val, 'foo')
+
+        try:
+            artifact3 = yield channel.next()
+            tools.assert_true(False, "Channel unexpectedly had an extra artifact")
```

Comment:
It would be even handier if you could ask whether a channel is done without having to call next() one more time.

Until then, I like the util.assert_channel_done(chan) idea.